File: week_1/day-5-string-hashing/variations_que/findAllAnagrams.js
function findAnagrams(s, p) {
    const res = [];
    const pCount = {};
    const window = {};

    for (let ch of p) pCount[ch] = (pCount[ch] || 0) + 1;

    for (let i = 0; i < s.length; i++) {
        window[s[i]] = (window[s[i]] || 0) + 1;

        if (i >= p.length) {
            const left = s[i - p.length];
            if (window[left] === 1) delete window[left];
            else window[left]--;
        }

        if (JSON.stringify(window) === JSON.stringify(pCount)) {
            res.push(i - p.length + 1);
        }
    }
    return res;
}

console.log(findAnagrams("cbaebabacd", "abc"));

File: week_1/day-5-string-hashing/variations_que/find_all_anagrams.py
def findAnagrams(s, p):
    from collections import Counter
    res = []
    p_count = Counter(p)
    window = Counter()

    for i, ch in enumerate(s):
        window[ch] += 1
        if i >= len(p):
            left_char = s[i - len(p)]
            if window[left_char] == 1:
                del window[left_char]
            else:
                window[left_char] -= 1

        if window == p_count:
            res.append(i - len(p) + 1)

    return res

# INPUT
print(findAnagrams("cbaebabacd", "abc"))  # [0, 6]
